lib/libxdp/xsk.c

```diff
@@ -0,0 +1,947 @@
+
+	err = bpf_map__resize(map, max_queues);
+	if (err)
+		return err;
+
+	return 0;
+}
+
+static void xsk_delete_map_entry(struct xsk_socket *xsk)
+{
+	struct xsk_ctx *ctx = xsk->ctx;
+
+	bpf_map_delete_elem(ctx->xsks_map_fd, &ctx->queue_id);
+	close(ctx->xsks_map_fd);
+}
+
+static int xsk_lookup_bpf_map(struct xsk_socket *xsk)
+{
+	__u32 i, *map_ids, num_maps, prog_len = sizeof(struct bpf_prog_info);
+	__u32 map_len = sizeof(struct bpf_map_info);
+	struct bpf_prog_info prog_info = {};
+	struct xsk_ctx *ctx = xsk->ctx;
+	struct bpf_map_info map_info;
+	int fd, err;
+
+	err = bpf_obj_get_info_by_fd(ctx->prog_fd, &prog_info, &prog_len);
+	if (err)
+		return err;
+
+	num_maps = prog_info.nr_map_ids;
+
+	map_ids = calloc(prog_info.nr_map_ids, sizeof(*map_ids));
+	if (!map_ids)
+		return -ENOMEM;
+
+	memset(&prog_info, 0, prog_len);
+	prog_info.nr_map_ids = num_maps;
+	prog_info.map_ids = (__u64)(unsigned long)map_ids;
+
+	err = bpf_obj_get_info_by_fd(ctx->prog_fd, &prog_info, &prog_len);
+	if (err)
+		goto out_map_ids;
+
+	ctx->xsks_map_fd = -1;
+
+	for (i = 0; i < prog_info.nr_map_ids; i++) {
+		fd = bpf_map_get_fd_by_id(map_ids[i]);
+		if (fd < 0)
+			continue;
+
+		err = bpf_obj_get_info_by_fd(fd, &map_info, &map_len);
+		if (err) {
+			close(fd);
+			continue;
+		}
+
+		if (!strcmp(map_info.name, "xsks_map")) {
+			ctx->xsks_map_fd = fd;
+			continue;
+		}
+
+		close(fd);
+	}
+
+	err = 0;
+	if (ctx->xsks_map_fd == -1)
+		err = -ENOENT;
+
+out_map_ids:
+	free(map_ids);
+	return err;
+}
+
+static int xsk_add_map_entry(struct xsk_socket *xsk)
+{
+	struct xsk_ctx *ctx = xsk->ctx;
+
+	return bpf_map_update_elem(ctx->xsks_map_fd, &ctx->queue_id,
+				   &xsk->fd, 0);
+}
+
+static int xsk_create_xsk_struct(int ifindex, struct xsk_socket *xsk)
+{
+	char ifname[IFNAMSIZ];
+	struct xsk_ctx *ctx;
+	char *interface;
+
+	ctx = calloc(1, sizeof(*ctx));
+	if (!ctx)
+		return -ENOMEM;
+
+	interface = if_indextoname(ifindex, &ifname[0]);
+	if (!interface) {
+		free(ctx);
+		return -errno;
+	}
+
+	ctx->ifindex = ifindex;
+	memcpy(ctx->ifname, ifname, IFNAMSIZ -1);
+	ctx->ifname[IFNAMSIZ - 1] = 0;
+
+	xsk->ctx = ctx;
+
+	return 0;
+}
+
+static enum xdp_attach_mode xsk_convert_xdp_flags(__u32 xdp_flags)
+{
+	if (xdp_flags & XDP_FLAGS_REPLACE)
+		pr_warn("XDP_FLAGS_REPLACE not supported by libxdp.\n");
+
+	if (xdp_flags & XDP_FLAGS_SKB_MODE)
+		return XDP_MODE_SKB;
+	if (xdp_flags & XDP_FLAGS_DRV_MODE)
+		return XDP_MODE_NATIVE;
+	if (xdp_flags & XDP_FLAGS_HW_MODE)
+		return XDP_MODE_HW;
+
+	return XDP_MODE_UNSPEC;
+}
+
+static int xsk_lookup_program(struct xsk_socket *xsk)
+{
+	struct xdp_multiprog *multi_prog;
+	struct xdp_program *prog = NULL;
+	struct xsk_ctx *ctx = xsk->ctx;
+
+	multi_prog = xdp_multiprog__get_from_ifindex(ctx->ifindex);
+	if (!multi_prog)
+		return -1;
+
+	while ((prog = xdp_multiprog__next_prog(prog, multi_prog)))
+		if (!strcmp(xdp_program__name(prog), "xsk_def_prog"))
+			break;
+
+	if (!prog)
+		return -1;
+	return xdp_program__fd(prog);
```

Comment:
This leaks the multi_prog reference... Given what I said above re: just keeping the xdp_program reference, maybe the right thing here is to have this function return an xdp_program pointer, and expose xdp_program__clone() so it can still close the multiprog...

Reply:
I will make a stab at just using the program pointer and make sure to destroy it when necessary.